libplatsupport/src/plat/polarfire/uart.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2020, DornerWorks
+ *
+ * This software may be distributed and modified according to the terms of
+ * the BSD 2-Clause license. Note that NO WARRANTY is provided.
+ * See "LICENSE_BSD2.txt" for details.
+ *
+ * @TAG(DORNERWORKS_BSD)
+ */
+
+#include <autoconf.h>
+
+#include <stdlib.h>
+#include <platsupport/serial.h>
+#include <platsupport/plat/serial.h>
+#include <platsupport/plat/icicle_mss.h>
+#include <string.h>
+
+#include "../../chardev.h"
+
+static inline uart_regs_t*
+uart_get_priv(ps_chardevice_t *d)
+{
+    return (uart_regs_t*)d->vaddr;
+}
+
+int uart_getchar(ps_chardevice_t *d)
+{
+    uart_regs_t* regs = uart_get_priv(d);
+    if (regs->line_status & LSR_DATA_READY_MASK)
+        return regs->rx_buffer;
+    return -1;
+}
+
+int uart_putchar(ps_chardevice_t* d, int c)
+{
+    uart_regs_t* regs = uart_get_priv(d);
+
+    while ((regs->line_status & LSR_TX_HOLD_REG_EMPTY_MASK) == 0);
+    regs->tx_buffer = c;
+    if (c == '\n' && (d->flags & SERIAL_AUTO_CR))
+    {
+        uart_putchar(d, '\r');
```

Comment:
This writes "\n\r" (LF+CR), but it should actually be "\r\n" (CR+LF ). That's a thing spreading through copy/paste since some time, see also https://github.com/seL4/util_libs/pull/30/commits).

I recommend to rewrite this as

```
static void busy_wait_fifo_empty_and_tx_char(uart_regs_t* regs, int c)
{
    // wait until FIFO empty
    while ((regs->line_status & LSR_TX_HOLD_REG_EMPTY_MASK) == 0)
    {
        // busy loop
    }

    regs->tx_buffer = c;
}

int uart_putchar(ps_chardevice_t* d, int c)
{
    uart_regs_t* regs = uart_get_priv(d);

    // turn LF into CR+LF if SERIAL_AUTO_CR is active
    if ((c == '\n') && (d->flags & SERIAL_AUTO_CR))
    {
        busy_wait_fifo_empty_and_tx_char(d,'\r');
    }

    busy_wait_fifo_empty_and_tx_char(d, c);
    return c;
}
```

Reply:
Nice catch, yeah I copied it from another driver.